File: adjacency_list/adjacency_list.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
    
edges = [('A', 'L'), ('A', 'N'), ('N', 'B'), ('B', 'N'), ('L', 'A'), ('L', 'B')]
nodes = ['A', 'N', 'B', 'L']

def build_adjacency_list(edges, nodes):
    adjacency_list = []

    for node in nodes:
        head = ListNode(node)
        adjacency_list.append(head)
        current = head
        for target in nodes:
            if (head.val, target) in edges:
                current.next = ListNode(target)
                current = current.next
    
    return adjacency_list

adjacency_list = build_adjacency_list(edges, nodes)

for list_node in adjacency_list:
    current = list_node
    while current:
        print(current.val, end=" -> " if current.next else "\n")
        current = current.next

File: adjacency_list/adjacency_list.go
package main

import "fmt"

type ListNode struct {
	Val  string
	Next *ListNode
}

func buildAdjacencyList(edges [][2]string, nodes []string) []*ListNode {
	edgeMap := make(map[[2]string]bool)
	for _, edge := range edges {
		edgeMap[edge] = true
	}

	var adjacencyList []*ListNode

	for _, nodeVal := range nodes {
		head := &ListNode{Val: nodeVal}
		adjacencyList = append(adjacencyList, head)
		current := head

		for _, target := range nodes {

			if edgeMap[[2]string{nodeVal, target}] {
				current.Next = &ListNode{Val: target}
				current = current.Next
			}
		}
	}

	return adjacencyList
}

func main() {
	edges := [][2]string{
		{"A", "L"}, {"A", "N"}, {"N", "B"},
		{"B", "N"}, {"L", "A"}, {"L", "B"},
	}
	nodes := []string{"A", "N", "B", "L"}

	adjacencyList := buildAdjacencyList(edges, nodes)

	for _, head := range adjacencyList {
		current := head
		for current != nil {
			if current.Next != nil {
				fmt.Printf("%s -> ", current.Val)
			} else {
				fmt.Printf("%s\n", current.Val)
			}
			current = current.Next
		}
	}
}
